src/generic.c

```diff
@@ -62,36 +71,45 @@ char *string_array_get_item(StringArray *arr, int index) {
 
 Blob *string_array_serialize(StringArray *arr) {
 	int message_size = 128;
-	int message_index = 0;
+	unsigned int message_index = 0;
 	Blob *blob = create_blob(message_size);
 
-	// first four bytes will be the number of strings in the array
-	*((int *)blob->bytes) = arr->length;
-	message_index += 4;
+	// write the wire header
+	struct ttetris_string_array_wire_header *header =
+	    (struct ttetris_string_array_wire_header *)blob->bytes;
+	header->max_string_length = arr->max_string_length;
+	header->length = arr->length;
+	message_index += sizeof(struct ttetris_string_array_wire_header);
 
 	// copy the strings into the BLOB
 	for (int i = 0; i < arr->length; i++) {
-		int len = strlen(arr->strings[i]) + 1;
+		unsigned int len =
+		    strnlen(arr->strings[i], arr->max_string_length) + 1;
 		// if the blob is not big enough, double it
 		while (message_index + len >= message_size) {
 			message_size *= 2;
 			resize_blob(blob, message_size);
 		}
-		memcpy(blob->bytes + message_index, arr->strings[i], len);
+		strncpy(blob->bytes + message_index, arr->strings[i], len);
 		message_index += len;
 	}
 
 	return blob;
 }
 
 StringArray *string_array_deserialize(Blob *blob) {
-	int length = *((int *)blob->bytes);
-	StringArray *arr = string_array_create(length);
-
-	int message_index = 4;
-
-	for (int i = 0; i < length; i++) {
-		int string_length = strlen(blob->bytes + message_index);
+	// read the header first
```

Comment:
Some of this is getting a little complex. It might not be a bad idea to document how bits/bytes are used in these blobs?

Reply:
Yeah; that's fair. I'll create an issue for documenting the wire format.

The `StringArray` serializer might be the most complex that exists in our code right now, tbh. My hope is that introducing the [struct ttetris_string_array_wire_header](https://github.com/Nitepone/terminally-tetris/blob/88a277429027f9bbede8ecf12556cdd087616809/src/generic.c#L27) in this patch at least clears up the first few bytes. I've started using structs more to increase readability around our wire format when practical.

Reply:
Created: #31